server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/SegmentSlotRelation.java

```diff
@@ -0,0 +1,57 @@
+package org.infinispan.server.resp.commands.cluster;
+
+import static org.infinispan.commons.logging.Log.CONFIG;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.infinispan.distribution.ch.impl.CRC16HashFunctionPartitioner;
+
+public class SegmentSlotRelation {
+
+   static final int SLOT_SIZE = 1 << 14;
+   private final int segmentSize;
+   private final int width;
+   private final Map<Integer, List<Integer>> relation;
+
+   public SegmentSlotRelation(int segmentSize) {
+      if (!CRC16HashFunctionPartitioner.isPow2(segmentSize))
+         throw CONFIG.respCacheSegmentSizePow2(segmentSize);
+      this.segmentSize = segmentSize;
+      this.width = 1 << Integer.numberOfTrailingZeros(SLOT_SIZE) - Integer.numberOfTrailingZeros(segmentSize);
+      Map<Integer, List<Integer>> relation = new HashMap<>(segmentSize);
+      for (int i = 0; i < SLOT_SIZE; i++) {
+         Integer k = i % segmentSize;
+         relation.computeIfAbsent(k, ignore -> new ArrayList<>(this.width)).add(i);
+      }
+
+      this.relation = Collections.unmodifiableMap(relation);
+   }
+
+   public int segmentToSingleSlot(int hash, int segment) {
```

Comment:
This method can be done just using math instead of having all these lookups which is going to be slower for sure.

If you use the block method I suggest in my previous comment it is just
`SLOT_SIZE/segmentSize * segment + reduce(hash)` iiuc. To be honest you may or may not want to store SLOT_SIZE/segmentSize as an instance variable.